### DISCUSSION TITLE
Idea to rewrite with Laravel + React

### DISCUSSION THREAD

**Discussion by yigitkerem** [server-status-project/server-status#164] [Ideas]
This project has a special place in my heart. Even though I believe I don't have the greatest of contributions, this is the first ever project I have contributed to. Seeing it stale like this is not the best of feelings. Looking at the project though, under today's standards, it looks awful, especially alongside a well designed website. Bootstrap itself is outdated as a design standard, yet we have an outdated version of it. Trying to find a place to clean it up yields no easy way, everything is now quite messed up and quite buggy, the development branches specifically. My idea is to rewrite with the following stack:
- Laravel
- Inertia.js
- React _(Open to suggestions on other SPA frameworks as well. Make sure it supports [inertia.js](https://intertiajs.com) though.)_
- TailwindCSS
This will give us very nice benefits:
- More stability
- Better looks
- More features
- Easy integration
- Better API
- Possibility of SaaS versions of Server-Status in the future :-) (just thinking out loud)
and more. 
I know it will need some time to be invested in it, but I am ready to for some time.

@Pryx @thnilsen @jhuesser @Nothing4You @kim3-sudo

**Comment by Tealk** [comment]
I have tried symfony
https://codeberg.org/RollenspielMonster/server-status-page

**Comment by yigitkerem** [comment]
I guess Laravel is a bit more simplified version of symfony.

**Reply by Tealk** [reply]
since i am not very familiar with either system, i cannot say which is easier.
Found symfony not so hard, actually only a database connection is missing that I can not get then that would even be ready.

**Comment by Pryx** [comment]
Hello everyone, 

so as some of you might know, I never expected this project to take off at all. It was originally a quick school project and as these projects tend to, it was built at the last minute before the deadline :smile: After finishing my studies, I tried to work on a bigger update, but got quickly discouraged, because of some of my design choices back then. I played with the idea of scraping the project and starting anew, but I never had enough time to go through with it. The furthest I got was a redesign mockup :smile: 

The key idea at the time was to make the status page as easy to set up and to make it compatible with as many hosting services as possible. And although it was achieved in most cases there were difficulties mainly with the mysqlnd driver not being present everywhere (a framework like Doctrine could easily fix this). Also one of the key requirements at the time was that the page was supposed to work just fine with Javascript turned off and that is one thing I would love to keep at least on the front page.

I will probably have some time for the project in the summer, so I would be happy to lead an organized effort at the complete rewrite. I would also be happy to finish up and adopt the Symfony version @Tealk posted. I have some experience with Symfony ecosystem so I think that direction would be good.

So what I would expect the rewrite to be like:

+ PHP 8.0+ (as php 7 active support ended last year)
+ Using type declarations and other features that were not originally present in PHP 5 :smile: 
+ Caching - I never had any performance issues, but caching the content would be a good idea.
+ Easy migration path from current version - as in make it at least possible to keep the incident history and services by recreating the config file and running some migrations.
+ Administration - make the editing experience much more streamlined. Could be written in React or some other library, interacting with an API.
+ API - Should have an OpenApi specification present and should be tested

I wouldn't necessarily focus on new functionality just yet, just get it as close as possible feature-wise, publish version 2.0 and then add new features :slightly_smiling_face:

**Reply by Tealk** [reply]
you are welcome to use this, if you are not comfortable with the frontend i can take a look at it; likes to create frontents

it would be great if a backend would be added that checks if a service is offline and then creates a post accordingly

**Comment by kim3-sudo** [comment]
For longer-term support, might I suggest using a modern framework like React or Node instead? If we're trying to get away from the issue of outdated frameworks, then IMHO, PHP is not the move to go for here. Even the new versions carry over a lot of the issues that plagued earlier PHP code. (I am a believer that the only reason PHP is still alive today is WordPress; if not for that, it should be a dead language.) If we're going for a full rewrite *anyways*, then React or Node are the modern web development tools that would ideally keep some level of compatibility with older versions. Just my two cents...

**Reply by Pryx** [reply]
I am pretty sure that there already are some node.js alternatives. The main idea was to be able to get a status page running on any of the many PHP hosting offers that would make it practically free. So even though I like node.js and rewriting it could be pretty cool, I don't really want to switch languages for that reason. React would be a nice addition though, mainly for the admin interface.

**Reply by Tealk** [reply]
i think the big advantage of php is simply that it runs on any webserver
